src/libstd/sys/unix/rwlock.rs

```diff
@@ -69,18 +69,21 @@ impl RWLock {
     #[inline]
     pub unsafe fn write(&self) {
         let r = libc::pthread_rwlock_wrlock(self.inner.get());
-        // See comments above for why we check for EDEADLK and write_locked. We
-        // also need to check that num_readers is 0.
+        // See comments above for why we check for EDEADLK and write_locked. For the same reason,
+        // we also need to check that there are no readers (tracked in `num_readers`).
         if r == libc::EDEADLK
-            || *self.write_locked.get()
+            || (r == 0 && *self.write_locked.get())
             || self.num_readers.load(Ordering::Relaxed) != 0
         {
+            // Above, we make sure to only access `write_locked` when `r == 0` to avoid
+            // data races.
             if r == 0 {
+                // `pthread_rwlock_wrlock` succeeded when it should not have.
                 self.raw_unlock();
             }
             panic!("rwlock write lock would result in deadlock");
         } else {
-            debug_assert_eq!(r, 0);
+            assert_eq!(r, 0);
```

Comment:
Why did you change this from a debug assert? The docs for `pthread_rwlock_wrlock` specify that it only ever returns `EDEADLK` or 0.

Reply:
Consistency -- I did the same in https://github.com/rust-lang/rust/pull/55865. Seems cheap to check.

But if you want I can also make it a debug assertion again.

Reply:
I would prefer making it a debug assert since this is a pretty hot path.

Reply:
Okay, I did that and added an appropriate comment.
Is it generally the case that the list of error codes in the POSIX docs is exhaustive? I noticed the docs explicitly say "These functions shall not return an error code of [EINTR]" but do not say that for any of the other error codes.

Reply:
They are supposed to be exhaustive in theory.